日報の言及機能を追加しました。

kayama3 opened the pull request:
確認よろしくお願いします〜

kayama3 committed:
追加：Mentionテーブルを追加

kayama3 committed:
追加：N:Nの自己参照の関係を設定

kayama3 committed:
修正：データが正しく保存できるようにDBを修正

Mentionに外部キー制約を追加、Mentionのリレーションを見直しました

kayama3 committed:
作成：メンションのview部分を作成

kayama3 committed:
作成：create時にメンションを作成

kayama3 committed:
追加：２重メンションを防止

DBのユニーク制約、モデルのバリデーション(uniqueness)を追加

kayama3 committed:
追加：update時の処理、ユニーク制約を追加

kayama3 committed:
修正：rubocopとerblintを通して修正

JunichiIto commented on app/models/mention.rb:
```diff
@@ -0,0 +1,9 @@
+  belongs_to :mentioned_report, class_name: 'Report'
+
+  validates :mentioning_report_id, presence: true, uniqueness: { scope: :mentioned_report_id }
+  validates :mentioned_report_id, presence: true, uniqueness: { scope: :mentioning_report_id }
```
belongs_toはデフォルトで入力必須になるので`presence: true`は省略可能です。
あとユニーク制約のバリデーションは同じ組み合わせであればどちらかひとつだけで大丈夫です。

JunichiIto commented on app/views/reports/show.html.erb:
```diff
@@ -6,6 +6,15 @@
+  <p><%= t('views.common.mention_from', name: Report.model_name.human) %></p>
+  <% @mentioned_reports.each do |report| %>
+    <ul>
+      <li><a href="http://localhost:3000/reports/<%= report[:id] %>"><%= report[:title] %></a></li>
```
link_to ヘルパーを活用した方がRailsらしいです。

JunichiIto commented on app/views/reports/show.html.erb:
```diff
@@ -6,6 +6,15 @@
+<% if @mentioned_reports.exists? %>
+  <p><%= t('views.common.mention_from', name: Report.model_name.human) %></p>
+  <% @mentioned_reports.each do |report| %>
+    <ul>
```
これだとレポートの件数分だけulタグが出力されるのでul/liの使い方としてあまり良くないです〜。

JunichiIto commented on app/controllers/reports_controller.rb:
```diff
@@ -9,6 +9,7 @@ def index
 
   def show
     @report = Report.find(params[:id])
+    @mentioned_reports = @report.mentioned_reports
```
以下の記事を読んでコントローラ全体の実装を見直してみてください〜。

[【新人プログラマ応援】Railsにおける良いコントローラ、悪いコントローラについて \- Qiita](https://qiita.com/jnchito/items/2b57316b39d92ca0e8dc)

JunichiIto reviewed (commented):
コメントしました！

kayama3 committed:
修正：指摘箇所の修正

kayama3 commented:
修正しましたので確認お願いします

kayama3 commented:
> 修正しましたので確認お願いします

https://github.com/kayama3/fjord-books_app-2023/blob/9b3f56cc4384d2e84e66c383129c92b4c41c006b/app/models/report.rb#L24-L25
https://github.com/kayama3/fjord-books_app-2023/blob/9b3f56cc4384d2e84e66c383129c92b4c41c006b/app/models/report.rb#L32-L33

kayama3 committed:
修正：重複する処理を修正

kayama3 commented:
修正したので確認お願いします。

JunichiIto commented on app/controllers/reports_controller.rb:
```diff
@@ -23,6 +23,8 @@ def create
     if @report.save
       redirect_to @report, notice: t('controllers.common.notice_create', name: Report.model_name.human)
+
+      @report.create_mentions
```
この方式でも結局「日報を保存したら言及も作成される」という処理がコントローラに限定されてしまい、rails consoleやバッチ処理で再利用できなくなりますね。なのでモデル内で完結する方が理想的です。

あと、`@report.save`した直後に何らかのトラブルが起きて`@report.create_mentions`が実行されなかった、ということが起きると、「日報は保存されたが、言及は作成されなかった」というデータの不整合が発生します。両方成功する or 両方失敗する（何もデータは作成されない）とするのが理想です。

そもそもの話ですが、`redirect_to`や`render`を呼んだ後にはなるべく他の処理を呼ばないようにするのがRails wayですね。

JunichiIto commented on app/views/reports/show.html.erb:
```diff
@@ -6,6 +6,15 @@
+<% if @report.mentioned_reports.exists? %>
+  <p><%= t('views.common.mention_from', name: Report.model_name.human) %></p>
+  <ul>
+    <% @report.mentioned_reports.each do |report| %>
```
並び順が一定になるようにorderを指定した方がいいと思います。

https://qiita.com/jnchito/items/90c28c3f3e856add0a82

JunichiIto reviewed (commented):
コメントしました！

kayama3 committed:
修正：処理をコントローラーに集約　@report.mentioned_reportsにorderを指定

kayama3 commented:
修正箇所の確認お願いします！

JunichiIto commented on app/models/report.rb:
```diff
@@ -25,13 +25,21 @@ def mentioning_report_ids
 
-  def create_mentions
+  def create_report_mentions
+    save
```
バリデーションエラーが起きて保存に失敗しても、無視されて先に進んでしまいそうです。

JunichiIto commented on app/models/report.rb:
```diff
@@ -25,13 +25,21 @@ def mentioning_report_ids
+    save
+
     mentioning_report_ids.each do |report_id|
       mentioning_relations.create(mentioned_report_id: report_id)
```
ここも上と同じく、バリデーションエラーが無視されそうです。

あと、mentioning_report_idsが10件あったとして、6件目で何らかのエラーが起きた場合、5件は保存に成功して、6件は失敗したまま、という問題が起きそうです。（update時も同様）

JunichiIto commented on app/models/report.rb:
```diff
@@ -25,13 +25,21 @@ def mentioning_report_ids
+    end
+
     unmentioned_report_ids = mentioning_reports.map(&:id) - mentioning_report_ids
     mentioning_relations.where(mentioned_report_id: unmentioned_report_ids).map(&:destroy)
```
```suggestion
    mentioning_relations.where(mentioned_report_id: unmentioned_report_ids).destroy_all
```

で同じことができます。

JunichiIto commented on app/models/report.rb:
```diff
@@ -25,13 +25,21 @@ def mentioning_report_ids
+      mentioning_relations.create(mentioned_report_id: report_id)
+    end
+
     unmentioned_report_ids = mentioning_reports.map(&:id) - mentioning_report_ids
```
1つのメソッド内で`content.scan(%r{http://localhost:3000/reports/(\d+)}).flatten.map(&:to_i)`が2回呼ばれるので、1回にした方が効率が良さそうだなと思いました。

JunichiIto commented on app/controllers/reports_controller.rb:
```diff
@@ -21,22 +21,16 @@ def edit; end
     @report = current_user.reports.new(report_params)
 
-    if @report.save
+    if @report.create_report_mentions
```
`create_report_mentions`は明示的に成功と失敗をtrue/falseで返す仕様になってますか？（updateも同様）

JunichiIto replied:
この件は未対応でしょうか？

JunichiIto reviewed (commented):
コメントしました〜。

kayama3 committed:
修正：指摘箇所の修正

create_report_mentionsとupdate_report_mentionsが明示的にtrue/falseを返すように修正
バリデーションエラーが無視されないように修正

kayama3 commented on app/models/report.rb:
```diff
@@ -14,4 +14,45 @@ def editable?(target_user)
+    transaction do
+      new_mentions = create_new_mentions(mentioned_report_ids)
+      new_mentions.each(&:save!)
+      mentioning_relations.where(mentioned_report_id: unmentioned_reports).map(&:destroy!)
```
この部分について、確かに`destroy_all`で同じような処理をすることはできるのですが、トランザクションのことを考えると例外が発生する`map(&:destroy!)`の形の方が良いかと思うのですが、どうでしょうか。

JunichiIto replied:
> トランザクションのことを考えると例外が発生する`map(&:destroy!)`の形の方が良い

ここは`destroy_all`と`map(&:destroy!)`で具体的にどういう違いがあって、後者が良いと考えたのか、詳しく教えてもらいたいですー。

あと、用途的にここで使うメソッドはmapではないように思います（理由を考えてみてください）。

kayama3 reviewed (commented):
確認よろしくお願いします〜

JunichiIto commented on app/models/report.rb:
```diff
@@ -14,4 +14,45 @@ def editable?(target_user)
+  has_many :mentioned_relations, class_name: 'Mention', foreign_key: :mentioned_report_id, dependent: :destroy, inverse_of: :mentioned_report
+  has_many :mentioned_reports, through: :mentioned_relations, source: :mentioning_report, dependent: :destroy
+
+  def find_report_ids
```
クラスの外部から呼ばれないメソッドはprivateメソッドにしましょう

JunichiIto commented on app/models/report.rb:
```diff
@@ -14,4 +14,45 @@ def editable?(target_user)
+
+    report_ids = find_report_ids
+
+    transaction do
```
これだと41行目のsave!でエラーが起きたときに、35行目のsave!だけが成功している、という状態が起きそうです。

JunichiIto commented on app/models/report.rb:
```diff
@@ -14,4 +14,45 @@ def editable?(target_user)
+  end
+
+  def create_report_mentions
+    return unless save!
```
save!だと、保存失敗時に例外が発生するので、そもそもunlessが条件分岐として機能するケースはないように思います。
（たとえばtitleを空にして保存しようとしたらどうなりますか？）

JunichiIto commented on app/models/report.rb:
```diff
@@ -14,4 +14,45 @@ def editable?(target_user)
+    report_ids = find_report_ids
+
+    transaction do
+      new_mentions = create_new_mentions(report_ids)
```
Railsで`create`というと、「新しいレコードをDBに保存する処理」を想像してしまいます。
newするだけならbuildを使うことが多いですね。
https://api.rubyonrails.org/classes/ActiveRecord/Associations/CollectionProxy.html#method-i-build

とはいえ、40-41行目はひとつのメソッドにまとめて問題ないんじゃないかなーと思います。

JunichiIto reviewed (commented):
コメントしました〜。

JunichiIto reviewed (commented):
コメントしました〜。

JunichiIto reviewed (commented):
コメントしました〜。

kayama3 committed:
修正：create_report_mentionsとupdate_report_mentionsが明示的にtrue/falseを返すように修正

kayama3 commented:
> ここはdestroy_allとmap(&:destroy!)で具体的にどういう違いがあって、後者が良いと考えたのか、詳しく教えてもらいたいですー。

まずは両者の違いから

- `destroy_all`は成功・失敗に関わらず、消去対象のインスタンスの配列を返す。
- `map(&:destroy!)`は成功時には消去したインスタンスの配列を、失敗時には`ActiveRecord::RecordNotDestroyed`を返す。

次にmap(&:destroy!)の方が良いと考えた理由ですが、私は

- destoy!は失敗時に例外を返すこと
- transaction内の処理で例外が起こるとrollbackすること

という２点に着目して、transaction内で`!`付きのメソッドを使うことで、失敗時にはrollbackしてデータの整合性を保証しようと考えました。
そのため、成功時には消去したインスタンスを、失敗時には`ActiveRecord::RecordNotDestroyed`を返す`destroy!`を使った`map(&:destroy!)`が良いと考えました。

kayama3 commented:
> あと、用途的にここで使うメソッドはmapではないように思います（理由を考えてみてください）。

各要素に対して処理をするだけの`each`に対して、`map`は戻り値として配列を返す。
処理をした後の配列を利用するのであれば`map`を用いるのが適切だが、今回の場合は日報に紐付いている言及を消去するだけで、処理をした後の配列を利用することはないので、`each`メソッドを用いる方が適切。
という理解でいいでしょうか？

kayama3 commented on app/models/report.rb:
```diff
@@ -14,4 +14,57 @@ def editable?(target_user)
+  has_many :mentioned_relations, class_name: 'Mention', foreign_key: :mentioned_report_id, dependent: :destroy, inverse_of: :mentioned_report
+  has_many :mentioned_reports, through: :mentioned_relations, source: :mentioning_report, dependent: :destroy
+
+  def create_report_mentions
```
### `!`付きのメソッドを使ってrollbackをする方針を変更したことについて(`update_report_mentions`も共通)
  - `!`付きのメソッドは失敗時に例外を返すことを利用して、処理の失敗時にrollbackを起こそうとしていたが、その部分を変更
  - [\[Rails\] Transaction内で条件分岐してrescue \- Qiita](https://qiita.com/aba0312/items/8b29cafe6b505b6a7a4f)より

### 処理の詳細
1. transaction内では`!`なしのメソッドを使用して処理を行う
    - 戻り値のtrue/falseを利用するため
2. `all_valid &= true/false`の形で、処理結果を変数に入れていく
3. `unless all_valid`の時に明示的にrollbackする
4. メソッドの最後にall_validを返すようにする
    - all_validの中身がfalseであれば、rollbackした上でfalseを、trueであればtrueをメソッドの返り値として返す

kayama3 commented on app/models/report.rb:
```diff
@@ -14,4 +14,57 @@ def editable?(target_user)
+      unmentioned_report_ids = mentioning_reports.map(&:id) - report_ids
+
+      all_valid &= create_new_mentions(mentioned_report_ids)
+      records = mentioning_relations.where(mentioned_report_id: unmentioned_report_ids).destroy_all
```
### map(&:destory!)をdestroy_allに変更したことについて
  - `!`付きのメソッドを使わなくなった以上、わざわざeach(&:destroy)の形にしなくてもいいかなと思いました
  - それに加えて、大量のレコードを消去する場合はdestroy_allを使用した方が効率がいいようなので、destroy_allを選択しました

kayama3 commented on app/models/report.rb:
```diff
@@ -14,4 +14,57 @@ def editable?(target_user)
+
+      all_valid &= create_new_mentions(mentioned_report_ids)
+      records = mentioning_relations.where(mentioned_report_id: unmentioned_report_ids).destroy_all
+      all_valid &= records.all?(&:destroyed?)
```
### .all(&:destroyed?)で全て消去されたかどうか確認
  - 消去が確実に行われたかを確認し、エラーが起こった場合にrollbackするためにこの部分を追加
  - [ActiveRecordにおけるdestroyとdestroy\!の違い \- Qiita](https://qiita.com/jnchito/items/3393c5c1a744199e128a#%E5%8F%82%E8%80%832-destroy_all%E3%82%92%E4%BD%BF%E3%81%86%E5%A0%B4%E5%90%88%E3%81%AE%E6%B3%A8%E6%84%8F%E7%82%B9)より
### 処理の詳細
1. recordsにdestroy_allした返り値を入れる
    - recordsには消去対象となったインスタンスの配列が入っている
2. `records.all?(&:destroyed?)`で全てのインスタンスが消去されているか確認し、全て消去されているのであればtrueを、そうでないのならfalseを返す

kayama3 replied:
### .all(&:destroyed?)で全て消去されたかどうか確認
  - 消去が確実に行われたかを確認し、エラーが起こった場合にrollbackするためにこの部分を追加
  - [ActiveRecordにおけるdestroyとdestroy\!の違い \- Qiita](https://qiita.com/jnchito/items/3393c5c1a744199e128a#%E5%8F%82%E8%80%832-destroy_all%E3%82%92%E4%BD%BF%E3%81%86%E5%A0%B4%E5%90%88%E3%81%AE%E6%B3%A8%E6%84%8F%E7%82%B9)より
### 処理の詳細
1. recordsにdestroy_allした返り値を入れる
    - recordsには消去対象となったインスタンスの配列が入っている
2. `records.all?(&:destroyed?)`で全てのインスタンスが消去されているか確認し、全て消去されているのであればtrueを、そうでないのならfalseを返す

JunichiIto commented on app/models/report.rb:
```diff
@@ -21,38 +21,50 @@ def created_on
+
+      report_ids = find_report_ids
+      mentioned_report_ids = report_ids - mentioning_reports.map(&:id)
+      unmentioned_report_ids = mentioning_reports.map(&:id) - report_ids
```
`mentioning_reports.map(&:id)`が2回出てくるのでDRYにしたいです。
あと、`mentioning_reports.pluck(:id)`や`mentioning_reports.ids`という書き方もあります。
こっちの方が少し効率の良いSQLが発行されます。

JunichiIto commented on app/models/report.rb:
```diff
@@ -21,38 +21,50 @@ def created_on
+      mentioning_relations.new(mentioned_report_id: report_id)
     end
+
+    new_mentions.each(&:save)
```
mapとeachでループが2回走るので、1回のループで終わらせたいなと思いました。

JunichiIto reviewed (commented):
コメントしました！

> 略... eachメソッドを用いる方が適切。という理解でいいでしょうか？

その理解で合ってます！ 🙆‍♂️

kayama3 committed:
修正：指摘箇所の修正　変数名を変更

JunichiIto commented on app/models/report.rb:
```diff
@@ -14,4 +14,56 @@ def editable?(target_user)
+    transaction do
+      all_valid &= update(params)
+
+      adding_repot_ids = find_adding_report_ids
```
```suggestion
      adding_report_ids = find_adding_report_ids
```

ですね

JunichiIto commented on app/models/report.rb:
```diff
@@ -14,4 +14,56 @@ def editable?(target_user)
+      mentioned_report_ids = adding_repot_ids - existing_report_ids
+      unmentioned_report_ids = existing_report_ids - adding_repot_ids
+
+      all_valid &= create_new_mentions(mentioned_report_ids)
```
create_new_mentions は常に真を返す設計になっているようです🧐

JunichiIto reviewed (commented):
コメントしました〜。

kayama3 committed:
修正：create_new_mentionsがtrue/falseを返すように修正

create_new_mentions内のeachをmapに変更
report.rb 49行目にall?を追加

JunichiIto commented on app/models/report.rb:
```diff
@@ -14,4 +14,56 @@ def editable?(target_user)
+      mentioned_report_ids = adding_report_ids - existing_report_ids
+      unmentioned_report_ids = existing_report_ids - adding_report_ids
+
+      all_valid &= create_new_mentions(mentioned_report_ids).all?
```
create_report_mentions でも`all?` を付ける必要がありますね。
そもそもの設計として、create_new_mentions が配列ではなく、true/falseを返すようになっていれば、`.all?`の付け忘れといううっかりミスを防止できそうです。

JunichiIto commented on app/views/reports/show.html.erb:
```diff
@@ -6,6 +6,15 @@
 <%= render 'shared/comments', commentable: @report %>
 
+<% if @report.mentioned_reports.exists? %>
+  <p><%= t('views.common.mention_from', name: Report.model_name.human) %></p>
```
この翻訳は特定の画面でしか使われないと思うので、commonではないと思います。

```
<%= t('.hogehoge') %>
```

のように、viewのパスに応じた翻訳キーを選ぶようにするといいかも？

JunichiIto reviewed (commented):
コメントしました〜。

kayama3 committed:
修正：指摘箇所の修正

create_new_mentionsがtrue/falseを返すように修正
reports/showの10行目翻訳をviewのパスに対応するように修正

JunichiIto commented on app/models/report.rb:
```diff
@@ -62,8 +62,10 @@ def find_adding_report_ids
 
   def create_new_mentions(ids)
-    ids.map do |report_id|
+    new_mentions = ids.map do |report_id|
```
map→all? ではなく、 直接 all? にブロックを渡す方がシンプルになりそうです。

JunichiIto reviewed (commented):
コメントしました〜。

kayama3 committed:
修正：create_new_mentions内のmapをall?に変更
